folium/plugins/search.py

```diff
@@ -2,83 +2,105 @@
         {% macro script(this, kwargs) %}
-
-            var {{this.get_name()}} = new L.GeoJSON({{this.data}});
-
-            {{this._parent.get_name()}}.addLayer({{this.get_name()}});
-
-            var searchControl = new L.Control.Search({
-                layer: {{this.get_name()}},
+            let {{this.layer.get_name()}}searchZoom = {{this.search_zoom}};
+            
+            var {{this.layer.get_name()}}searchControl = new L.Control.Search({
+                layer: {{this.layer.get_name()}},
                 propertyName: '{{this.search_label}}',
+                collapsed: {{this.collapsed}},
+                textPlaceholder: '{{this.placeholder}}',
             {% if this.geom_type == 'Point' %}
                 initial: false,
-                zoom: {{this.search_zoom}},
+                zoom: {{this.layer.get_name()}}searchZoom?{{this.layer.get_name()}}searchZoom:{{this.parent_map_name}}.getZoom(),
                 position:'{{this.position}}',
                 hideMarkerOnCollapse: true
-            {% endif %}
-            {% if this.geom_type == 'Polygon' %}
+            {% else %}
                 marker: false,
                 moveToLocation: function(latlng, title, map) {
-                var zoom = {{this._parent.get_name()}}.getBoundsZoom(latlng.layer.getBounds());
-                    {{this._parent.get_name()}}.setView(latlng, zoom); // access the zoom
+                var zoom = {{this.layer.get_name()}}searchZoom?{{this.layer.get_name()}}searchZoom:map.getBoundsZoom(latlng.layer.getBounds())
+                    map.setView(latlng, zoom); // access the zoom
                 }
             {% endif %}
                 });
-                searchControl.on('search:locationfound', function(e) {
-
-                    e.layer.setStyle({fillColor: '#3f0', color: '#0f0'});
+                {{this.layer.get_name()}}searchControl.on('search:locationfound', function(e) {
+                    {{this.layer.get_name()}}.setStyle(function(feature){
+                        return feature.properties.style
+                    })
+                    {% if this.options %}
+                    e.layer.setStyle({{ this.options }});
+                    {% endif %}
                     if(e.layer._popup)
                         e.layer.openPopup();
-
-                }).on('search:collapsed', function(e) {
-
-                    {{this.get_name()}}.eachLayer(function(layer) {   //restore feature color
-                        {{this.get_name()}}.resetStyle(layer);
+                })
+                {{this.layer.get_name()}}searchControl.on('search:collapsed', function(e) {
+                        {{this.layer.get_name()}}.setStyle(function(feature){
+                            return feature.properties.style
                     });
                 });
-            {{this._parent.get_name()}}.addControl( searchControl );
+            {{this.parent_map_name}}.addControl( {{this.layer.get_name()}}searchControl );
```

Comment:
Seems we no longer need the whole `parent_map_name` stuff. You already assert `Search` is added to `Map`, so you can get its parent directly. I did like the function you made for it, but it doesn't make sense to keep unused code. Removing it will probably also solve the conflict.

Reply:
Agreed. Will update shortly.